### DISCUSSION TITLE
gensound creates temp audio files which don't get clean up afterwards

### DISCUSSION THREAD

**Discussion by kodcdangky** [Quefumas/gensound#28] [Q&A] [answered]
Hello there,

Really loving this project, every feature I have tried so far has been an absolute wonder!

I have only 1 problem so far that I hope someone can help me fix. 

```
import time
import gensound as gs

audio = gs.WAV('D:\\System\\Desktop\\water_alarm.mp3')
audio.play()
time.sleep(2)
```

When I run this code on PyCharm, it plays normally leaving nothing behind (or at least none that I can find anywhere). But if I export this code using pyinstaller into an .exe file and run that, a folder called gensound_temp is created where I have the .exe file, and a .wav temp file is generated every time play() is called. This folder and all its files are kept permanently, or until I manually delete them. Is it possible for these files to be cleaned up afterwards?

If this is unwanted behavior, please let me know what info can I provide to help diagnose the problem.
I currently have gensound 0.5.2, numpy 1.22.1, pygame 2.1.2, pyinstaller 4.8 installed, if that helps.

**Comment by Quefumas** [comment]
Welcome, and thanks for the feedback!

My suggestions:
1. If all you're doing is playing from already-existing WAV/MP3 files, I suggest not using gensound at all, since it would be much simpler with PyGame or any other alternatives. I can help navigate those, if needed.
2. It's possible (maybe?) that due to pyinstaller, gensound can't find PyGame and then resorts to more primitive I/O capabilities. See if you can run `from gensound.io import IO; IO.status(True)` in both environments, and copy the output. This should tell you which I/O capabilities gensound is aware of in both environments, and we can test this hypothesis and see what we can do.
3. Try to load a WAV rather than MP3 file (you can use the temporary one for example), and see if playback creates another new file or not.
4. Try to load an MP3, but instead of `play()`, use `export('filename.wav')`, and see if this creates any redundant wave files (other than the desired output).

Steps 3&4 will let us know if the temporary file is created because of playback limitations, or due to reading MP3 files, which are not supported natively by Python. If step 3 does not create an extra file, a temporary solution would be to just use WAVs (gensound can help convert), although that's far from ideal of course. Another non-ideal solution is to call `IO.cleanup()` before exiting the application, which attempts to clear these temporary files, with varying amounts of success.

**Comment by Quefumas** [comment]
And two more things:

1. if the application plays the same file/signal more than once, this is much more efficient:
```python
wav = gs.WAV('D:\\System\\Desktop\\water_alarm.mp3')
audio = wav.realise() # create Audio object once
audio.play() # run this as many times as you want
time.sleep(2)
```
This is because every time you run `play()` on the original `WAV` object, this reprocesses the entire audio stream. This way, the processing is separated from the playback (see more in the [IO Guide](https://github.com/Quefumas/gensound/wiki/IO-Guide)). Perhaps I should make it so that calling `play` on a pure `WAV` object with no modifications should skip processing and play directly from file.

2. Use `time.sleep(2)` with some care. It will act differently according to the playback interface used. I think PyGame plays the audio in the background (though maybe we can pass an argument to change that), but some other interfaces will pause until playback ends.

**Reply by kodcdangky** [reply]
Hey, thanks for the 2 quick responses!

The code i quoted was just for testing purpose, hence the sleep since pygame plays sounds asynchronously. My actual project is at https://github.com/kodcdangky/Periodic_Timer which also changes the sound volume (and maybe more in the future) before playing the sound which is why I was looking for a proper audio library. (it's really messy code though, sorry about that, I'm mostly a beginner)

About your suggestions, I tried this and unfortunately it still created 3 temp files for the 3 .play() called
```
import time
import gensound as gs

audio = gs.WAV('D:\\System\\Desktop\\water_alarm.wav')
audio.realise(sample_rate=44100)

for _ in range(3):
    audio.play()
    time.sleep(0.5)
```

I'll try running IO.status(True) next, but currently my workaround is using pydub for modifying audio and pygame.mixer for playing (since pydub doesn't support playing sound async) and after exporting with pyinstaller it still plays normally so I don't think this is the problem, but I'll try nonetheless

Thanks so much for your help!

**Reply by Quefumas** [reply]
No problem!

Note that you have to keep the result of `audio.realise` - it outputs an `Audio` object, which contains the signal after processing. The code above calls the `play` method for the same object as in the original, that's why there's no difference (notice line 2 in my code above).

Having said that, and looking at your code, here is an option that I imagine should work:
```python
import time
import gensound as gs
from gensound.io import _IO_pygame

audio = gs.WAV('D:\\System\\Desktop\\water_alarm.wav')
processed_audio = audio.realise(sample_rate=44100)

_IO_pygame.playback(processed_audio)
```

**Reply by kodcdangky** [reply]
Sorry about that, I should've looked more carefully.

Unfortunately, keeping the result of `audio.realise` then playing didn't change that it still created a temp file for every `play()` called

Also, calling `_IO_pygame.playback(realised_audio)` gave me an `AttributeError` because my audio is missing `byte_width`, which while I can see that it's attached to my file if i call `play()`, I don't know how to do that by myself.

Your theory in suggestion #2 was right! In PyCharm my `IO.status(True)` result is `playback: pygame (simpleaudio, playsound, winsound, os)`, but exported version is only `playback: winsound (os)`, and importing pygame to the test .exe didn't change that either, which really makes me wonder how my workaround is working

Thanks again!

p.s. Unrelated to gensound so you really don't have to answer, but with my workaround, every time `mixer.Sound(audio).play()` is called in the .exe file, a split-second console window popped on. Is there a way to suppress this on `pygame.mixer`?

**Reply by Quefumas** [reply]
Thanks, you're raising a lot of good issues here.

Before I start getting wordy, a fix on my suggestion:
```python
from gensound.io import _IO_pygame

audio = gs.WAV('D:\\System\\Desktop\\water_alarm.wav')
processed_audio = audio.realise(sample_rate=44100)
processed_audio._prepare_buffer(2, None) # <----- added this
_IO_pygame.playback(processed_audio)
```

**Reply by Quefumas** [reply]
Wait! I think this should be working too:

```python
import gensound as gs
from gensound.io import IO

IO.set_io("play", "pygame") # call this once; this forces gensound to use pygame

gs.WAV('D:\\System\\Desktop\\water_alarm.wav').play()
```

The method `set_io` was meant to switch between available interfaces, but I don't see why it shouldn't work even for devices that it doesn't know are available.

**Reply by Quefumas** [reply]
The boring part:
The reason your workaround was working was that pygame was always available (importable), but the mechanism that gensound uses for detecting which interfaces are available is not working well inside a pyinstaller environment (not surprising). Gensound could tell what's available by trying to import it, but that would take time and fill up memory for absolutely no good reason. 

This mechanism is lacking in other areas too, but if using `set_io` works for this, that's already something.

**Reply by kodcdangky** [reply]
`IO.set_io` worked, and for .mp3 as well as .wav, thank you so much!
Also I always like to try to understand the underlying mechanism, so no worries
Cheers! And thanks again

**Reply by Quefumas** [reply]
Thank you for letting me know about this. I'll try to figure out how to fix this in future versions, and now at least we know of a workaround.

(On an aside, you asked about console window opening - I haven't really used any of PyInstaller or PyGame, so I can't help you much, but I would try to see whether using any other feature of PyGame has the same effect. Perhaps PyGame is used to running in a console by default, so it creates one for itself)

**Comment by kodcdangky** [comment, marked as answer]
https://github.com/Quefumas/gensound/discussions/28#discussioncomment-2068837 <- The answer followed by the explanation
